Error when trying to consume a topic with NULL values

When running a local test in my dev schema after upgrading the Kafka Connector to 1.0.0, I ran into the following issue:

```
[main] INFO org.apache.kafka.clients.consumer.internals.Fetcher - [Consumer clientId=consumer-EXASOL_KAFKA_UDFS_CONSUMERS-1, groupId=EXASOL_KAFKA_UDFS_CONSUMERS] Fetch position FetchPosition{offset=117013, offsetEpoch=Optional.empty, currentLeader=LeaderAndEpoch{leader=Optional[<IP>:<PORT> (id: 22 rack: null)], epoch=0}} is out of range for partition public.pro-seller.offer-0, resetting offset
[main] INFO org.apache.kafka.clients.consumer.internals.SubscriptionState - [Consumer clientId=consumer-EXASOL_KAFKA_UDFS_CONSUMERS-1, groupId=EXASOL_KAFKA_UDFS_CONSUMERS] Resetting offset for partition public.pro-seller.offer-0 to position FetchPosition{offset=139566, offsetEpoch=Optional.empty, currentLeader=LeaderAndEpoch{leader=Optional[<IP>:<PORT> (id: 22 rack: null)], epoch=0}}.
[main] INFO org.apache.kafka.common.metrics.Metrics - Metrics scheduler closed
[main] INFO org.apache.kafka.common.metrics.Metrics - Closing reporter org.apache.kafka.common.metrics.JmxReporter
[main] INFO org.apache.kafka.common.metrics.Metrics - Metrics reporters closed
[main] INFO org.apache.kafka.common.utils.AppInfoParser - App info kafka.consumer for consumer-EXASOL_KAFKA_UDFS_CONSUMERS-1 unregistered
F-UDF-CL-SL-JAVA-1080: Exception during run
com.exasol.cloudetl.kafka.KafkaConnectorException: Error consuming Kafka topic 'public.pro-seller.offer' data. It occurs for partition '0' in node '0' and vm '140681706548736' Cause: null
com.exasol.cloudetl.kafka.KafkaTopicDataImporter$.run(KafkaTopicDataImporter.scala:99)
com.exasol.cloudetl.kafka.KafkaTopicDataImporter.run(KafkaTopicDataImporter.scala)
com.exasol.ExaWrapper.run(ExaWrapper.java:196)
Caused by: java.lang.NullPointerException
com.exasol.cloudetl.kafka.KafkaTopicDataImporter$.$anonfun$run$1(KafkaTopicDataImporter.scala:84)
com.exasol.cloudetl.kafka.KafkaTopicDataImporter$.$anonfun$run$1$adapted(KafkaTopicDataImporter.scala:71)
scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
scala.collection.AbstractIterable.foreach(Iterable.scala:919)
com.exasol.cloudetl.kafka.KafkaTopicDataImporter$.run(KafkaTopicDataImporter.scala:71)
	... 6 more
 
W-UDF.CL.SL.JAVA-1082: Skipping init, because cleanup method cannot be found.
```
This seems to relate to the following [line in KafkaTopicDataImporter.scala](https://github.com/exasol/kafka-connector-extension/blob/47c28892f896183794f912f8234d30c4eff61f5f/src/main/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporter.scala#L84):

```scala
          val exasolRow: Seq[Any] = recordValue ++ metadata
```

Is there a possibility that the above construct will fail if the `recordValue` is null? This is something that quite frequently happens in the topic I'm reading here, as a way to indicate for a record to be deleted..

Previously, that particular line looked like:

```scala
          val exasolRow: Seq[Object] = getAvroRow(singleColumnJson, recordValue) ++ metadata
```

..where `getAvroRow()` was defined like:

```scala
  private def getAvroRow(singleJson: Boolean, recordValue: GenericData.Record): Seq[Object] =
    if (singleJson) {
      Seq(s"$recordValue")
    } else {
      AvroRow(recordValue).getValues().map(_.asInstanceOf[AnyRef])
    }
```

Apparently the null conversion would pass by just fine as part of the string interpolation above (I'm extracting value as JSON).

--- a/.travis.yml
+++ b/.travis.yml
@@ -21,10 +21,10 @@ before_install:
 matrix:
   include:
     - jdk: oraclejdk11
-      scala: 2.13.5
+      scala: 2.13.6
 
     - jdk: openjdk11
-      scala: 2.13.5
+      scala: 2.13.6
 
 script:
   - ./scripts/ci.sh

--- a/build.sbt
+++ b/build.sbt
@@ -10,8 +10,8 @@ lazy val orgSettings = Seq(
 )
 
 lazy val buildSettings = Seq(
-  scalaVersion := "2.13.5",
-  crossScalaVersions := Seq("2.12.13", "2.13.5")
+  scalaVersion := "2.13.6",
+  crossScalaVersions := Seq("2.12.14", "2.13.6")
 )
 
 lazy val root =

--- a/doc/changes/changelog.md
+++ b/doc/changes/changelog.md
@@ -1,5 +1,6 @@
 # Releases
 
+* [1.1.0](changes_1.1.0.md)
 * [1.0.0](changes_1.0.0.md)
 * [0.2.1](changes_0.2.1.md)
 * [0.2.0](changes_0.2.0.md)

--- a/doc/changes/changes_1.1.0.md
+++ b/doc/changes/changes_1.1.0.md
@@ -1,0 +1,31 @@
+# Kafka Connector Extension 1.1.0, released 2021-06-XX
+
+Code name: 
+
+## Summary
+
+## Features
+
+## Bug
+
+* #29: Fixed issues with record null values in a topic
+
+### Runtime Dependency Updates
+
+* Updated `sbt.version:1.4.4` to `1.5.3`
+* Updated Scala version `2.13.5` to `2.13.6`
+* Updated `org.apache.kafka:kafka-clients:2.6.0` to `2.8.0`
+* Updated `org.scala-lang.modules:scala-collection-compat:2.4.3` to `2.4.4`
+
+### Test Dependency Updates
+
+* Updated `org.scalatest:scalatest:3.2.8` to `3.2.9`
+* Updated `org.mockito:mockito-core:3.9.0` to `3.11.0`
+
+### Plugin Updates
+
+* Updated `com.eed3si9n:sbt-assembly:0.15.0` to `1.0.0`
+* Updated `com.typesafe.sbt:sbt-git:1.0.0` to `1.0.1`
+* Updated `org.scoverage:sbt-scoverage:1.7.3` to `1.8.2`
+* Updated `org.wartremover:sbt-wartremover:2.4.13` to `2.4.15`
+* Updated `org.wartremover:sbt-wartremover-contib:1.3.11` to `1.3.12`

--- a/project/Dependencies.scala
+++ b/project/Dependencies.scala
@@ -10,12 +10,12 @@ object Dependencies {
   private val ImportExportUDFVersion = "0.2.0"
   private val KafkaClientsVersion = "2.8.0"
   private val KafkaAvroSerializerVersion = "6.1.1"
-  private val ScalaCollectionCompatVersion = "2.4.3"
+  private val ScalaCollectionCompatVersion = "2.4.4"
 
   // Test dependencies versions
-  private val ScalaTestVersion = "3.2.8"
+  private val ScalaTestVersion = "3.2.9"
   private val ScalaTestPlusVersion = "1.0.0-M2"
-  private val MockitoCoreVersion = "3.9.0"
+  private val MockitoCoreVersion = "3.11.0"
   private val KafkaSchemaRegistryVersion = "6.1.1"
 
   val Resolvers: Seq[Resolver] = Seq(

--- a/project/Settings.scala
+++ b/project/Settings.scala
@@ -36,7 +36,6 @@ object Settings {
     // ScalaFmt settings
     scalafmtOnCompile := true,
     // Scoverage settings
-    coverageMinimum := 50,
     coverageOutputHTML := true,
     coverageOutputXML := true,
     coverageOutputCobertura := true,

--- a/project/build.properties
+++ b/project/build.properties
@@ -1,1 +1,1 @@
-sbt.version=1.4.4
+sbt.version=1.5.3

--- a/project/plugins.sbt
+++ b/project/plugins.sbt
@@ -1,15 +1,15 @@
 // Adds a `wartremover` a flexible Scala code linting tool
 // http://github.com/puffnfresh/wartremover
-addSbtPlugin("org.wartremover" % "sbt-wartremover" % "2.4.13")
+addSbtPlugin("org.wartremover" % "sbt-wartremover" % "2.4.15")
 
 // Adds Contrib Warts
 // http://github.com/wartremover/wartremover-contrib/
-addSbtPlugin("org.wartremover" % "sbt-wartremover-contrib" % "1.3.11")
+addSbtPlugin("org.wartremover" % "sbt-wartremover-contrib" % "1.3.12")
 
 // Adds a `assembly` task to create a fat JAR with all of its
 // dependencies
 // https://github.com/sbt/sbt-assembly
-addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.15.0")
+addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "1.0.0")
 
 // Adds most common doc api mappings
 // https://github.com/ThoughtWorksInc/sbt-api-mappings
@@ -17,7 +17,7 @@ addSbtPlugin("com.thoughtworks.sbt-api-mappings" % "sbt-api-mappings" % "3.0.0")
 
 // Adds Scala Code Coverage (Scoverage) used during unit tests
 // http://github.com/scoverage/sbt-scoverage
-addSbtPlugin("org.scoverage" % "sbt-scoverage" % "1.7.3")
+addSbtPlugin("org.scoverage" % "sbt-scoverage" % "1.8.2")
 
 // Adds SBT Coveralls plugin for uploading Scala code coverage to
 // https://coveralls.io
@@ -43,7 +43,7 @@ addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.10.0-RC1")
 
 // Adds a `git` plugin
 // https://github.com/sbt/sbt-git
-addSbtPlugin("com.typesafe.sbt" % "sbt-git" % "1.0.0")
+addSbtPlugin("com.typesafe.sbt" % "sbt-git" % "1.0.1")
 
 // Adds a `sbt-explicit-dependencies` plugin
 // https://github.com/cb372/sbt-explicit-dependencies

--- a/sbtx
+++ b/sbtx
@@ -34,11 +34,11 @@
 
 set -o pipefail
 
-declare -r sbt_release_version="1.4.4"
-declare -r sbt_unreleased_version="1.4.4"
+declare -r sbt_release_version="1.5.3"
+declare -r sbt_unreleased_version="1.5.3"
 
-declare -r latest_213="2.13.4"
-declare -r latest_212="2.12.12"
+declare -r latest_213="2.13.6"
+declare -r latest_212="2.12.14"
 declare -r latest_211="2.11.12"
 declare -r latest_210="2.10.7"
 declare -r latest_29="2.9.3"
@@ -48,7 +48,7 @@ declare -r buildProps="project/build.properties"
 
 declare -r sbt_launch_ivy_release_repo="https://repo.typesafe.com/typesafe/ivy-releases"
 declare -r sbt_launch_ivy_snapshot_repo="https://repo.scala-sbt.org/scalasbt/ivy-snapshots"
-declare -r sbt_launch_mvn_release_repo="https://repo.scala-sbt.org/scalasbt/maven-releases"
+declare -r sbt_launch_mvn_release_repo="https://repo1.maven.org/maven2"
 declare -r sbt_launch_mvn_snapshot_repo="https://repo.scala-sbt.org/scalasbt/maven-snapshots"
 
 declare -r default_jvm_opts_common="-Xms512m -Xss2m -XX:MaxInlineLevel=18"
@@ -247,11 +247,18 @@ java_version() {
   echo "$version"
 }
 
+is_apple_silicon() { [[ "$(uname -s)" == "Darwin" && "$(uname -m)" == "arm64" ]]; }
+
 # MaxPermSize critical on pre-8 JVMs but incurs noisy warning on 8+
 default_jvm_opts() {
   local -r v="$(java_version)"
   if [[ $v -ge 10 ]]; then
-    echo "$default_jvm_opts_common -XX:+UnlockExperimentalVMOptions -XX:+UseJVMCICompiler"
+    if is_apple_silicon; then
+      # As of Dec 2020, JVM for Apple Silicon (M1) doesn't support JVMCI
+      echo "$default_jvm_opts_common"
+    else
+      echo "$default_jvm_opts_common -XX:+UnlockExperimentalVMOptions -XX:+UseJVMCICompiler"
+    fi
   elif [[ $v -ge 8 ]]; then
     echo "$default_jvm_opts_common"
   else

--- a/scripts/ci.sh
+++ b/scripts/ci.sh
@@ -6,7 +6,7 @@ set -o errtrace -o nounset -o pipefail -o errexit
 BASE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )"/.. && pwd )"
 cd "$BASE_DIR"
 
-MAIN_SCALA_VERSION=2.13.5
+MAIN_SCALA_VERSION=2.13.6
 
 if [[ -z "${TRAVIS_SCALA_VERSION:-}" ]]; then
   echo "Environment variable TRAVIS_SCALA_VERSION is not set"

--- a/src/it/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporterNullValueIT.scala
+++ b/src/it/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporterNullValueIT.scala
@@ -35,12 +35,12 @@ class KafkaTopicDataImporterNullValueIT extends KafkaIntegrationTest {
 
     val iter = mockExasolIterator(properties, Seq(0), Seq(-1))
     val meta = mock[ExaMetadata]
-    when(meta.getOutputColumnCount()).thenReturn(3)
+    when(meta.getOutputColumnCount()).thenReturn(4)
     KafkaTopicDataImporter.run(meta, iter)
 
     verify(iter, times(2)).emit(Seq(any[Object]): _*)
-    verify(iter, times(1)).emit(null, JInt.valueOf(0), JLong.valueOf(0))
-    verify(iter, times(1)).emit(null, JInt.valueOf(0), JLong.valueOf(1))
+    verify(iter, times(1)).emit(null, null, JInt.valueOf(0), JLong.valueOf(0))
+    verify(iter, times(1)).emit(null, null, JInt.valueOf(0), JLong.valueOf(1))
   }
 
 }

--- a/src/it/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporterNullValueIT.scala
+++ b/src/it/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporterNullValueIT.scala
@@ -1,0 +1,46 @@
+package com.exasol.cloudetl.kafka
+
+import java.lang.{Integer => JInt}
+import java.lang.{Long => JLong}
+
+import scala.jdk.CollectionConverters._
+
+import com.exasol.ExaDataTypeException
+import com.exasol.ExaMetadata
+import com.exasol.cloudetl.kafka.KafkaTopicDataImporterAvroIT.schemaRegistryUrl
+
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.serialization.Serializer
+import org.apache.kafka.common.serialization.StringSerializer
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.mockito.ArgumentMatchers._
+import org.mockito.Mockito.{times, verify, when}
+
+class KafkaTopicDataImporterNullValueIT extends KafkaIntegrationTest {
+  override def additionalProperties: Map[String, String] =
+    Map("SCHEMA_REGISTRY_URL" -> schemaRegistryUrl)
+
+  implicit val serializer: Serializer[String] = {
+    val properties = Map("schema.registry.url" -> schemaRegistryUrl)
+    val serializer = new StringSerializer()
+    serializer.configure(properties.asJava, false)
+    serializer
+  }
+
+  test("run emit fills columns with null if value is null") {
+    createCustomTopic(topic)
+    implicit val serializer = new StringSerializer
+    publishToKafka(new ProducerRecord(topic, null: String))
+    publishToKafka(new ProducerRecord(topic, null: String))
+
+    val iter = mockExasolIterator(properties, Seq(0), Seq(-1))
+    val meta = mock[ExaMetadata]
+    when(meta.getOutputColumnCount()).thenReturn(3)
+    KafkaTopicDataImporter.run(meta, iter)
+
+    verify(iter, times(2)).emit(Seq(any[Object]): _*)
+    verify(iter, times(1)).emit(null, JInt.valueOf(0), JLong.valueOf(0))
+    verify(iter, times(1)).emit(null, JInt.valueOf(0), JLong.valueOf(1))
+  }
+
+}

--- a/src/it/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporterNullValueIT.scala
+++ b/src/it/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporterNullValueIT.scala
@@ -27,7 +27,7 @@ class KafkaTopicDataImporterNullValueIT extends KafkaIntegrationTest {
     serializer
   }
 
-  test("run emit fills columns with null if value is null") {
+  test("emit fills columns with null if record value is null") {
     createCustomTopic(topic)
     implicit val serializer = new StringSerializer
     publishToKafka(new ProducerRecord(topic, null: String))

--- a/src/main/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporter.scala
+++ b/src/main/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporter.scala
@@ -81,10 +81,11 @@ object KafkaTopicDataImporter extends LazyLogging {
           )
 
           val recordValue =
-            if (record.value() != null)
+            if (record.value() != null) {
               record.value()
-            else
+            } else {
               Seq.fill[Any](metadata.getOutputColumnCount().toInt - kafkaMetadata.length) { null }
+            }
 
           val exasolRow: Seq[Any] = recordValue ++ kafkaMetadata
           iterator.emit(exasolRow: _*)

--- a/src/main/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporter.scala
+++ b/src/main/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporter.scala
@@ -75,13 +75,18 @@ object KafkaTopicDataImporter extends LazyLogging {
               s"value '${record.value()}'"
           )
 
-          val metadata: Seq[Object] = Seq(
+          val kafkaMetadata: Seq[Object] = Seq(
             record.partition().asInstanceOf[AnyRef],
             record.offset().asInstanceOf[AnyRef]
           )
 
-          val recordValue = record.value()
-          val exasolRow: Seq[Any] = recordValue ++ metadata
+          val recordValue =
+            if (record.value() != null)
+              record.value()
+            else
+              Seq.fill[Any](metadata.getOutputColumnCount().toInt - kafkaMetadata.length) { null }
+
+          val exasolRow: Seq[Any] = recordValue ++ kafkaMetadata
           iterator.emit(exasolRow: _*)
 
         }

--- a/src/main/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporter.scala
+++ b/src/main/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporter.scala
@@ -10,6 +10,7 @@ import com.exasol.ExaMetadata
 import com.exasol.cloudetl.kafka.deserialization.{AvroDeserialization, JsonDeserialization}
 
 import com.typesafe.scalalogging.LazyLogging
+import org.apache.kafka.clients.consumer.ConsumerRecord
 import org.apache.kafka.common.TopicPartition
 
 /**
@@ -74,22 +75,7 @@ object KafkaTopicDataImporter extends LazyLogging {
               s"'${record.offset()}' with key '${record.key()}' and " +
               s"value '${record.value()}'"
           )
-
-          val kafkaMetadata: Seq[Object] = Seq(
-            record.partition().asInstanceOf[AnyRef],
-            record.offset().asInstanceOf[AnyRef]
-          )
-
-          val recordValue =
-            if (record.value() != null) {
-              record.value()
-            } else {
-              Seq.fill[Any](metadata.getOutputColumnCount().toInt - kafkaMetadata.length) { null }
-            }
-
-          val exasolRow: Seq[Any] = recordValue ++ kafkaMetadata
-          iterator.emit(exasolRow: _*)
-
+          iterator.emit(getRecordRow(record, metadata.getOutputColumnCount().toInt): _*)
         }
         logger.info(
           s"Emitted total '$totalRecordCount' records for partition " +
@@ -107,6 +93,20 @@ object KafkaTopicDataImporter extends LazyLogging {
     } finally {
       kafkaConsumer.close();
     }
+  }
+
+  private[this] def getRecordRow(
+    record: ConsumerRecord[String, Seq[Any]],
+    outputColumnCount: Int
+  ): Seq[Any] = {
+    val metadataColumns =
+      Seq(record.partition().asInstanceOf[AnyRef], record.offset().asInstanceOf[AnyRef])
+    val valueColumns = if (record.value() != null) {
+      record.value()
+    } else {
+      Seq.fill[Any](outputColumnCount - metadataColumns.length) { null }
+    }
+    valueColumns ++ metadataColumns
   }
 
 }

